### PR TITLE
feat(cli): support async nuxt.config.js

### LINKED ISSUE #4005: Async data inside nuxt.config.js
### What problem does this feature solve?

Currently there is no nice way how to use async data inside nuxt.config.js. 
For example getting some module configuration from remote server, getting head data, list of plugins that need to be enabled for current build etc.

This would allow to use same Nuxt app for multiple similar sites but with slight difference of configuration that is changeable from remote admin panel.

### What does the proposed changes look like?

Something like this (quite pseudocode) 
```js
// nuxt.config.js

const getHeadConfig = async () => {
  let { data } = await axios.get('/my/head/config/url')
  return {
    ...data,
    title: 'hurrdurr'
  }
}

module.exports = {
  mode: 'universal',

  // Get conf from remote with axios
  loading: axios.get('/my/loading/config/url'),

  // Use some method to get data from remote
  head: getHeadConfig()
}

```

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c7860">#c7860</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by TheAlexLichter**
## Types of changes
- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

With this PR, the `nuxt.config.js` export can be `async` now and will be `await`ed properly.
Resolves #4005 

## Checklist:
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #TODO)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by manniL** [comment]
This seemed a little bit too easy though :thinking:

**Review by galvez — DISMISSED** [review]
@manniL lovely :)

File: bin/nuxt-build [hunk 1]
```diff
@@ -40,49 +40,51 @@ if (argv.help) {
...
+      .then(close)
+      .catch(err => consola.fatal(err))
+  } else {
+    // Build + Generate for static deployment
+    new Generator(nuxt, builder)
+      .generate({ build: true })
+      .then(close)
+      .catch(err => consola.fatal(err))
+  }
+})()
```
Comment (pi0) [inline bin/nuxt-build]: The result of this promise is unhandled. We can call it like this:

```js
build()
  .then(() => { process.exit(0) } )
  .catch(err => { consola.fatal(err); process.exit(1) })
```

File: bin/common/utils.js [hunk 2]
```diff
@@ -49,6 +48,15 @@ exports.loadNuxtConfig = (argv) => {
     if (options.default) {
       options = options.default
     }
+
+    if (typeof options === 'function') {
+      try {
+        options = await options()
```
Comment (pi0) [inline bin/common/utils.js]: Usage of top-level `async/await` will break `nuxt-legacy` for Node 6 support. We have to decide to deprecate that package or use normal `Promise`.

File: bin/nuxt-build (hunk 1 above)
Comment (manniL) [reply bin/nuxt-build]: Right! I guess same goes for the other bin scripts?

Comment (pi0) [reply bin/nuxt-build]: Yep :) Another idea is to create a wrapper util to refactor this logic. Something like `runScript`.

Impl:

```js
async function runScript(fn) {
  try {
   await fn();
   process.exit(0);
  } catch(e) {
consola.fatal(err)
consola.fatal(err)
 }
}
```

File: bin/common/utils.js (hunk 2 above)
Comment (pi0) [reply bin/common/utils.js]: Alternative idea: Use webpack to transpile `nuxt-legacy/bin` scripts :D

Comment (manniL) [reply bin/common/utils.js]: Though all major cloud platforms support Node 8, I wouldn't like to see `nuxt-legacy` deprecated yet. So either Promises (which hurt the readability a little) or webpack (which introduces another build step, but only for the core-team and not for the users).

Webpack seems to be the better choice, doesn't it?

File: bin/nuxt-build (hunk 1 above)
Comment (manniL) [reply bin/nuxt-build]: But there is the top-level async problem again :see_no_evil:

Comment (manniL) [reply bin/nuxt-build]: Ah, no. We can avoid it with promises as well.

File: bin/common/utils.js (hunk 2 above)
Comment (pi0) [reply bin/common/utils.js]: Using babel or babylon to this is as simple as something like this:

```
yarn add --dev babel-cli
yarn babel bin -x ".js," -d packages/nuxt-legacy/bin
```

File: bin/common/utils.js [hunk 3]
```diff
@@ -31,6 +31,15 @@ const getLatestHost = (argv) => {
   return { port, host, socket }
 }
 
+exports.runAsyncScript = (fn) => {
+  fn().then(() => { process.exit(0) })
+    .catch((err) => {
+      consola.fatal('Failed to run async Buxt script!')
```
Comment (pi0) [inline bin/common/utils.js]: Typo + Script is not always build :)

File: bin/common/utils.js [hunk 4]
```diff
@@ -32,9 +31,20 @@ const getLatestHost = (argv) => {
...
 }
 
+exports.runAsyncScript = async (fn) => {
+  try {
+    await fn()
+    process.exit(0)
+  } catch (err) {
+    consola.fatal('Failed to run async Nuxt script!')
+    consola.fatal(err)
+    process.exit(1)
```
Comment (aldarund) [inline bin/common/utils.js]: Is it required? Is even second fatal executed? There is a hook to exit on fatal
https://github.com/nuxt/nuxt.js/blob/dev/bin/nuxt#L12-L19

Comment (manniL) [reply bin/common/utils.js]: Good catch!

**Review by pi0 — DISMISSED** [review]
Except for yarn.lock, LGTM 👍

**Comment by manniL** [comment]
Tests fixed! cc @pi0 @clarkdo

**Comment by manniL** [comment]
@pi0 Updated to latest dev ☺️

**Review by pi0 — APPROVED** [review]

**Pull request merged**